### PR TITLE
add grpc endpoint to retrieve oom events

### LINKED ISSUE #751: Add GRPC endpoint to retrieve OOM events
This issue is to create a GRPC endpoint that will allow the runtime/shim to retrieve OOM events from the agent.

This is the corresponding issue in the runtime: https://github.com/kata-containers/runtime/issues/2150

### PR DESCRIPTION
**Pull request by awprice**
fixes #751

Signed-off-by: Alex Price <aprice@atlassian.com>

### CODE REVIEW

**Comment by awprice** [comment]
I've raised this as a draft as I would like to get some feedback/thoughts on this approach before finishing/adding tests.

cc @lifupan @gnawux @devimc @amshinde @egernst

File: protocols/grpc/agent.proto [hunk 1]
```diff
@@ -60,6 +60,7 @@ service AgentService {
 	rpc MemHotplugByProbe(MemHotplugByProbeRequest) returns (google.protobuf.Empty);
 	rpc SetGuestDateTime(SetGuestDateTimeRequest) returns (google.protobuf.Empty);
 	rpc CopyFile(CopyFileRequest) returns (google.protobuf.Empty);
+	rpc GetOOMEvents(GetOOMEventsRequests) returns (OOMEvents);
```
Comment (awprice) [inline protocols/grpc/agent.proto]: The runtime/shim will poll this endpoint periodically to retrieve OOM events that have occurred since the last time it was polled. I haven't determined the poll period yet, but potentially something low like 500ms to ensure timely arrival of events.

Ideally this endpoint would be a stream endpoint, but ideally they shouldn't be used - explained here https://github.com/kata-containers/runtime/issues/2150#issuecomment-586346977

File: agent.go [hunk 2]
```diff
@@ -1492,6 +1503,7 @@ func realMain() error {
 		deviceWatchers: make(map[string](chan string)),
 		storages:       make(map[string]*sandboxStorage),
 		stopServer:     make(chan struct{}),
+		oomEvents:      make(chan string, 128),
```
Comment (awprice) [inline agent.go]: This channel will aggregate all of the oom event channels from the containers into a single buffered channel. A call to `GetOOMEvents` will read off all of the current events off the channel.

File: agent.go [hunk 3]
```diff
@@ -787,6 +788,16 @@ func (s *sandbox) listenToUdevEvents() {
 	}
 }
 
+func (s *sandbox) runOOMEventMonitor(ch <-chan struct{}, containerID string) {
+	go func() {
+		for {
+			<-ch
```
Comment (devimc) [inline agent.go]: not sure, but I think we are leaking routines here, what happen when the container is restarted? `runOOMEventMonitor` is called again and a new routine for the new container is started, but what happen with the old routines ?

**Comment by grahamwhaley** [comment]
Is there any chance/opportunity to add unit tests here to cover the new code?

File: agent.go (hunk 3 above)
Comment (lifupan) [reply agent.go]: Yes, I think: first , the container's oomCh could be closed when the container terminated,
then check the channel, if it's closed, break thhis loop and exit this go routine.

File: protocols/grpc/agent.proto (hunk 1 above)
Comment (lifupan) [reply protocols/grpc/agent.proto]: Maybe you can use a sync call to wait until there is OOM events occur and then return the events, published the events to containerd and then another call again in a loop.

File: grpc.go [hunk 4]
```diff
@@ -1791,3 +1798,13 @@ func (a *agentGRPC) StopTracing(ctx context.Context, req *pb.StopTracingRequest)
 
 	return emptyResp, nil
 }
+
+func (a *agentGRPC) GetOOMEvents(ctx context.Context, req *pb.GetOOMEventsRequests) (*pb.OOMEvents, error) {
+	capacity := len(a.sandbox.oomEvents)
```
Comment (lifupan) [inline grpc.go]: Is it better to wait here until there is an OOM events occurred and then return this call?

File: agent.go (hunk 3 above)
Comment (awprice) [reply agent.go]: Yep I can add logic to see if the channel is closed and then break the routine.

File: grpc.go [hunk 5]
```diff
@@ -861,6 +861,13 @@ func (a *agentGRPC) StartContainer(ctx context.Context, req *pb.StartContainerRe
 		return emptyResp, err
 	}
 
+	// Add the container to the OOM event monitor
+	oomCh, err := ctr.container.NotifyOOM()
+	if err != nil {
+		return emptyResp, err
+	}
+	a.sandbox.runOOMEventMonitor(oomCh, req.ContainerId)
```
[the remark this answers is not in the document]
Comment (awprice) [reply grpc.go]: Yep that's correct

File: agent.go [hunk 6]
```diff
@@ -787,6 +788,16 @@ func (s *sandbox) listenToUdevEvents() {
...
+			agentLog.WithField("container_id", containerID).Info("Received OOM event")
+			s.oomEvents <- containerID
```
[the remark this answers is not in the document]
Comment (awprice) [reply agent.go]: Is the lock to ensure that there is only 1 routine per container?

**Comment by awprice** [comment]
> Is there any chance/opportunity to add unit tests here to cover the new code?

Yep I plan on adding unit tests, I wanted to get some feedback on the design before spending the time writing tests.

File: agent.go (hunk 3 above)
Comment (awprice) [reply agent.go]: I've updated it to check if the channel is closed, and if so it will break the routine. This should prevent leaking of goroutines if the channel is closed.

File: grpc.go (hunk 4 above)
Comment (awprice) [reply grpc.go]: Yep I can do that if desired. In this case, would we only return a single event at a time?

Comment (lifupan) [reply grpc.go]: Yeah, I think one event a time can make the api much simple.

File: agent.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (devimc) [reply agent.go]: nop, the `s.Lock()` is to avoid race conditions between routines that access to `ooEvents`

File: grpc.go (hunk 4 above)
Comment (awprice) [reply grpc.go]: Updated GRPC endpoint to only return a single event at a time.

File: protocols/grpc/agent.proto (hunk 1 above)
Comment (awprice) [reply protocols/grpc/agent.proto]: I've updated the GRPC endpoint to return a single event at a time.

File: agent.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (awprice) [reply agent.go]: I've updated how the GRPC endpoint returns events - only a single at a time and will wait until there is an event ready to then return. The issue with using a `s.Lock()` now is that if nothing calls the GRPC endpoint, then in theory the routine will block waiting for something to remove it from the `oomEvents` channel and never unlock - potentially causing other issues.

[the remark this answers is not in the document]
Comment (devimc) [reply agent.go]: thanks @awprice - please add a unit test for  this function

[the remark this answers is not in the document]
Comment (awprice) [reply agent.go]: Thanks @devimc, I'll focus on adding unit tests next.

**Comment by awprice** [comment]
Thanks for the reviews everyone. I'll focus on adding unit tests now.

**Comment by awprice** [comment]
Some basic tests added.

**Comment by jcvenegas** [comment]
@awprice  hi this PR requires a rebase

File: agent.go [hunk 7]
```diff
@@ -787,6 +788,19 @@ func (s *sandbox) listenToUdevEvents() {
...
 }
 
+func (s *sandbox) runOOMEventMonitor(ch <-chan struct{}, containerID string) {
+	go func() {
+		for {
+			_, ok := <-ch
+			if !ok {
+				return
+			}
+			agentLog.WithField("container_id", containerID).Info("Received OOM event")
```
[the remark this answers is not in the document]
Comment (awprice) [reply agent.go]: Fixed!

**Comment by awprice** [comment]
> @awprice hi this PR requires a rebase

Sorry for the delay - fixed!

**Comment by awprice** [comment]
Tests are now failing - 
```
INFO: Building github.com/golangci/golangci-lint
go build -o golangci-lint ./cmd/golangci-lint
go: github.com/go-critic/go-critic@v0.0.0-20181204210945-ee9bf5809ead: invalid pseudo-version: does not match version-control timestamp (2019-02-10T22:04:43Z)
Makefile:8: recipe for target 'build' failed
make: *** [build] Error 1
The command ".ci/static-checks.sh" failed and exited with 2 during .
```
😢

**Comment by jodh-intel** [comment]
Urggh.. that might be fallout from https://github.com/kata-containers/tests/pull/1427 - @teawater - have you seen this issue before?

**Comment by grahamwhaley** [comment]
I restarted the Travis in the vain hope that maybe this has 'self healed' in the last few days - worth a shot to know where we are at I think...

**Comment by awprice** [comment]
Still failing at the same point :(

**Comment by grahamwhaley** [comment]
@chavafg - any ideas here? Given afaik travis is working for other new PRs, I have to wonder if:
 - it is for this repo only
 - or if maybe a rebase and repush will trigger travis afresh (maybe travis is being too smart and cacheing stuff - we've seen that before....)

**Comment by jodh-intel** [comment]
I'm not clear why we're only seeing this issue here. I've tried various versions of Ubuntu and golang but I cannot recreate this issue locally.

The problem (I *think*) is caused by a combination of go 1.13+ coupled with a package that uses go modules and which has dependencies that use go modules. For us, the problem looks to be being caused by [`golangci-lint`](https://github.com/golangci/golangci-lint) which uses modules as does the dependent [`go-critic`](https://github.com/go-critic/go-critic). 

# TL;DR

I think the solution should be to just bump the version of `golangci-lint` in the runtime's `versions.yaml` to the latest `v1.24.0` release (we're currently using `v1.15.0`).

**Comment by chavafg** [comment]
`golangci-lint` issue is now fixed, travis now reported a couple of `go test` failures
```
INFO: Running 'go test' as current user on package 'github.com/kata-containers/agent/protocols/client' with flags '-v -race -timeout 30s'
=== RUN   TestNewAgentClient
--- FAIL: TestNewAgentClient (1.01s)
	Error Trace:	client_test.go:114
			client_test.go:124
			client_test.go:134
	Error:      	Should be true
	Test:       	TestNewAgentClient
	Messages:   	expect err message: Invalid vsock port	got: parse vsock://100:bar: invalid port ":bar" after host
=== RUN   TestNewAgentClientWithYamux
--- FAIL: TestNewAgentClientWithYamux (1.04s)
	Error Trace:	client_test.go:114
			client_test.go:148
			client_test.go:156
	Error:      	Should be true
	Test:       	TestNewAgentClientWithYamux
	Messages:   	expect err message: Invalid vsock port	got: parse vsock://100:bar: invalid port ":bar" after host
=== RUN   TestParseGrpcHybridVSockAddr
--- PASS: TestParseGrpcHybridVSockAddr (0.00s)
FAIL
coverage: 62.7% of statements
FAIL	github.com/kata-containers/agent/protocols/client	2.065s
FAIL
Makefile:131: recipe for target 'test' failed
```

**Comment by awprice** [comment]
@chavafg Weird, I don't think I modified either of those tests or the functionality related to them. Has the version of Go being used changed?

**Comment by awprice** [comment]
@chavafg @jodh-intel Are you able to look at the failing test and update me on the next steps?

I'm pretty sure it's now failing due to a change in Go version used in the tests and the error message has now changed.
```
INFO: Running 'go test' as current user on package 'github.com/kata-containers/agent/protocols/client' with flags '-v -race -timeout 30s'
=== RUN   TestNewAgentClient
--- FAIL: TestNewAgentClient (1.01s)
	Error Trace:	client_test.go:114
			client_test.go:124
			client_test.go:134
	Error:      	Should be true
	Test:       	TestNewAgentClient
	Messages:   	expect err message: Invalid vsock port	got: parse vsock://100:bar: invalid port ":bar" after host
=== RUN   TestNewAgentClientWithYamux
--- FAIL: TestNewAgentClientWithYamux (1.04s)
	Error Trace:	client_test.go:114
			client_test.go:148
			client_test.go:156
	Error:      	Should be true
	Test:       	TestNewAgentClientWithYamux
	Messages:   	expect err message: Invalid vsock port	got: parse vsock://100:bar: invalid port ":bar" after host
=== RUN   TestParseGrpcHybridVSockAddr
--- PASS: TestParseGrpcHybridVSockAddr (0.00s)
FAIL
coverage: 62.7% of statements
FAIL	github.com/kata-containers/agent/protocols/client	2.065s
FAIL
Makefile:131: recipe for target 'test' failed
make: *** [test] Error 1
```

**Comment by fidencio** [comment]
It's been fixed already: https://github.com/kata-containers/agent/pull/773

**Comment by jodh-intel** [comment]
Restarted SLES CI which failed with:

```
not ok 1 measured time for /dev/random
# (in test file integration/entropy/entropy_time.bats, line 32)
#   `[ "$measured_time" -le "$expected_time" ]' failed
#
```

Restarted Travis which failed on these two tests:

```
=== RUN   TestNewAgentClient
--- FAIL: TestNewAgentClient (1.01s)
    assertions.go:247: 
                          
	Error Trace:	client_test.go:114
        
	            	
			client_test.go:124
        
	            	
			client_test.go:134
        
	Error:      	Should be true
        
	Test:       	TestNewAgentClient
        
	Messages:   	expect err message: Invalid vsock port	got: parse vsock://100:bar: invalid port ":bar" after host
=== RUN   TestNewAgentClientWithYamux
--- FAIL: TestNewAgentClientWithYamux (1.04s)
    assertions.go:247: 
                          
	Error Trace:	client_test.go:114
        
	            	
			client_test.go:148
        
	            	
			client_test.go:156
        
	Error:      	Should be true
        
	Test:       	TestNewAgentClientWithYamux
        
	Messages:   	expect err message: Invalid vsock port	got: parse vsock://100:bar: invalid port ":bar" after host
```

**Comment by chavafg** [comment]
weird, travis passed on x86 platform but not on ppc64... @awprice can you please rebase with latest changes? That should fix travis.

**Comment by awprice** [comment]
Thanks everyone for the reviews - is this good to merge? I'll get started on the runtime side of this.

**Comment by bergwolf** [comment]
Let's merge it Thanks @awprice !
